README.md
# mydb-expose

  Connect middleware to expose documents through MyDB.


History.md

3.0.0 / 2017-01-04
==================

 * Breaking API change: `monk` is no longer supported, this package
   now uses the mongodb driver directly.

2.1.2 / 2015-03-31
==================

 * session: copy reference to new cookie object

2.1.1 / 2015-03-28
==================

 * session: handle case where session field is replaced

2.1.0 / 2014-01-26
==================

 * docs
 * expose: add support for passing a mydb server ref as `url`

2.0.7 / 2013-08-29
==================

 * session: express csrf warning workaround

2.0.6 / 2013-08-15
==================

  * session: save more information in session

2.0.5 / 2013-08-14
==================

  * remove agent

2.0.4 / 2013-08-14
==================

  * expose: fix agent setting

2.0.3 / 2013-08-14
==================

  * expose: use an agent

2.0.2 / 2013-08-14
==================

  * expose: track subscription times

2.0.1 / 2013-06-14
==================

  * expose: handle socket errors

2.0.0 / 2013-06-14
==================

  * switch to RESTful subscription, ditch redis

1.1.8 / 2013-04-23
==================

  * expose: make mydb-expose session independent

1.1.7 / 2013-04-15
==================

  * expose: fix nasty state bug thanks to @visionmedia

1.1.6 / 2013-04-14
==================

  * expose: handle error

1.1.5 / 2013-04-14
==================

  * expose: make uid generation asynchronous

1.1.4 / 2013-04-04
==================

  * expose: fix fields retrieval from promise
  * test: for non-GET

1.1.3 / 2013-04-04
==================

  * expose: added `GET` check for `/session`

1.1.2 / 2013-03-07
==================

  * session: introduce `user_agent` in session

1.1.1 / 2013-02-21
==================

  * expose: consider socketid in hash

1.1.0 / 2013-02-21
==================

  * test: added `res#subscribe` tests
  * expose: added `toFields` helper
  * expose: expose `res.subscribe` and `req.subscribe`
  * expose: rename `doSubscripe` to `createSubscription`
  * expose: implemented `res.subscribe`
  * send: use `res.subscribe`
  * index: fix variable shadowing

1.0.1 / 2013-02-20
==================

  * expose: fix race condition
  * test: jesus christ redis

1.0.0 / 2013-02-19
==================

  * expose: fix publish callback
  * expose: improve subscription ids
  * expose: fix error reporting
  * package: added `uid2` dep
  * expose: prevent potential thrown JSON circular structure error
  * expose: create a new random `SocketId` if one is not present
  * expose: `doSubscribe` now performs a REDIS publish instead
  * expose: refactored `subscribe`
  * expose: added support for 404s for non-mydb requests
  * expose: handle error from `subscribe` method
  * epxose: use `MyDB` instead
  * expose: make check for mydb request based on presence of `X-MyDB-SocketId` header
  * expose: style
  * expose: added public subscribe method
  * expose: added new response mechanism

0.6.7 / 2013-02-05
==================

  * expose: added fix for circular JSON

0.6.6 / 2013-01-25 
==================

  * tweak Session#JSON() to return sess data only

0.6.5 / 2013-01-25 
==================

  * add Session#toJSON() to prevent cyclic ref

0.6.4 / 2013-01-25 
==================

  * add more debug() instrumentation

0.6.3 / 2013-01-09
==================

  * expose: allow session-less exposing

0.6.2 / 2012-11-08
==================

  * expose: fix error handling in `res.send` queries

0.6.1 / 2012-10-24
==================

  * expose: consider fields in hashing

0.6.0 / 2012-10-17
==================

  * index: added `sessionExpose` opt
  * expose: implement custom session fields

0.5.0 / 2012-10-16
==================

  * session: add `addToSet` method

0.4.2 / 2012-10-14
==================

  * session: maintain properties from `req.session`

0.4.1 / 2012-10-13
==================

  * expose: prevent including `mydb-expose` twice

0.4.0 / 2012-10-09
==================

  * expose: hash collection and id with sessionID

0.3.0 / 2012-10-09
==================

  * expose: send 404 when promise resolves to nothing
  * Makefile: tweak `slow`

0.2.1 / 2012-10-07
==================

  * index: fixed default mongodb uri

0.2.0 / 2012-10-06
==================

  * expose: added collection name support to subscriptions

0.1.0 / 2012-10-03
==================

  * mydb-expose: initial release


test/session.js

/**
 * Module dependencies.
 */

var Session = require('../session')
  , expect = require('expect.js');

describe('Session#toJSON()', function(){
  it('should return a JSON representation', function(){
    var obj = { first: 'tobi', last: 'ferret' };
    var req = { session: obj, req: req };
    var sess = new Session('users', req);

    expect(sess.toJSON()).to.eql({
      first: 'tobi',
      last: 'ferret'
    });
  })
})


test/mydb-expose.js
"use strict";

/**
 * Test dependencies.
 */

const mydbExpose = require('..');
const mydb = require('mydb');
const http = require('http');
const express = require('express');
const expect = require('expect.js');
const request = require('supertest');
const mongodb = require('mongodb');
const expressSession = require('express-session');

/**
 * Test.
 */

describe('mydb-expose', function(){
  let db, users, woots, sessions;

  /**
   * Middleware helper for tests.
   *
   * @api private
   */

  function expose(){
    return mydbExpose(db, { url: mydb(http.createServer()) });
  }

  /**
   * Returns the session middleware.
   *
   * @api private
   */

  function session(){
    return expressSession({ secret: 'woot', resave: false, saveUninitialized: true });
  }
  
  before(function (done) {
    mongodb.MongoClient
      .connect('mongodb://localhost/mydb-expose-test')
      .then(_db => {
        db = _db;
        users = db.collection('users-' + Date.now());
        woots = db.collection('woots-' + Date.now());
        sessions = db.collection('sessions');    
        done();
      })
      .catch(done);
  });
  
  describe('res#send', function(){
    
    let doc1, doc2;
    
    before(function(done) {
      doc1 = { _id: new mongodb.ObjectID(), tobi: 'rox' };
      doc2 = { _id: new mongodb.ObjectID(), jane: 'too' };

      Promise
        .all([users.insert(doc1), users.insert(doc2)])
        .then(() => {
          done();
        })
        .catch(done);
    })

    it('Collection#findOne', function(done){
      var app = express();
      app.use(session());
      app.use(expose());
      app.get('/doc', function(req, res){
        res.send(users.findOne(doc1._id));
      });
      request(app).get('/doc').expect({
        _id: doc1._id.toString(),
        tobi: doc1.tobi
      }, done);
    });

    it('Collection#find', function(done){
      var app = express();
      app.use(session());
      app.use(expose());
      app.get('/document', function(req, res){
        res.send(users.find({}));
      });
      request(app).get('/document').expect([
        { _id: doc1._id.toString(), tobi: doc1.tobi },
        { _id: doc2._id.toString(), jane: doc2.jane }
      ], done);
    });

    it('Collection#find + 404', function(done){
      var app = express();
      app.use(session());
      app.use(expose());
      app.get('/missing-doc', function(req, res){
        res.send(users.findOne({ asd: 'testing testing 404' }));
      });
      request(app)
      .get('/missing-doc')
      .end(function(err, res){
        if (err) return done(err);
        expect(res.status).to.be(404);
        done();
      });
    });
  });

  describe('/session', function(){
    it('responds json', function(done){
      var app = express();
      app.use(session());
      app.use(expose());
      request(app).get('/session').end(function(err, res){
        if (err) return done(err);

        // we first assert a session doc was created
        sessions.findOne({ _id: new mongodb.ObjectId(res.body._id) }, function(err, sess){
          
          if (err) return done(err);
          expect(sess.sid).to.be.a('string');

          request(app)
          .get('/session')
          .set('Cookie', res.headers['set-cookie'][0].split(';')[0])
          .end(function(err, res){
            if (err) return done(err);
            expect(res.body._id).to.be(sess._id.toString());
            expect(res.body.sid).to.be(undefined);
            done();
          });
        });
      });
    });

    it('ignores non-GET', function(done){
      var app = express();
      app.use(session());
      app.use(expose());
      request(app).post('/session').end(function(err, res){
        if (err) return done(err);
        expect(res.status).to.be(404);
        done();
      });
    });

    it('responds with a mydb id', function(done){
      var app = express();
      app.use(session());
      app.use(expose());
      app.get('/', function(req, res, next){
        res.send(req.sessionID);
      });

      request(app).get('/').end(function(err, res){
        if (err) return done(err);
        var sid = res.text;

        request(app)
        .get('/session?my=1')
        .set('X-MyDB-SocketId', 'woot')
        .set('Cookie', res.headers['set-cookie'][0].split(';')[0])
        .end(function(err, res){
          if (err) return done(err);
          expect(res.body._id).to.be.a('string');
          expect(res.headers['x-mydb-id']).to.be.a('string');
          done();
        });
      });
    });

    it('responds with 304', function(done){
      var app = express();
      app.use(session());
      app.use(expose());
      app.get('/', function(req, res, next){
        res.send(req.sessionID);
      });

      request(app)
      .get('/')
      .end(function(err, res){
        if (err) return done(err);
        var sid = res.text;
        var cookie = res.headers['set-cookie'][0].split(';')[0];

        request(app)
        .get('/session?my=1')
        .set('X-MyDB-SocketId', 'woot')
        .set('Cookie', cookie)
        .end(function(err, res){
          if (err) return done(err);
          expect(res.body._id).to.be.a('string');
          var id = res.headers['x-mydb-id'];

          request(app)
          .get('/session?my=1')
          .set('X-MyDB-Id', id)
          .set('X-MyDB-SocketId', 'woot')
          .set('Cookie', cookie)
          .end(function(err, res){
            expect(err).to.be(null);
            expect(res.status).to.be(304);
            done();
          });
        });
      });
    });
  });

  describe('req#session', function(){
    it('automatically populated', function(done){
      var app = express();
      app.use(session());
      app.use(expose());
      app.get('/', function(req, res, next){
        expect(req.session.sid).to.eql(req.sessionID);
        res.send('' + req.session._id);
      });
      app.get('/2', function(req, res, next){
        res.send('' + req.session._id);
      });
      request(app)
      .get('/')
      .expect(200)
      .end(function(err, res1){
        if (err) return done(err);
        var cookie = res1.headers['set-cookie'][0].split(';')[0];

        request(app)
        .get('/2')
        .set('Cookie', cookie)
        .expect(200)
        .end(function(err, res2){
          expect(res1.text).to.be(res2.text);
          done();
        });
      });
    });

    it('supports operations', function(done){
      var app = express();
      app.use(session());
      app.use(expose());

      app.get('/', function(req, res, next){
        res.sendStatus(200);
      });
      app.get('/2', function(req, res, next){
        req.session.set('woot', 'a');
        req.session.push('likes', 'ferrets');
        res.sendStatus(200);
      });
      app.get('/3', function(req, res, next){
        expect(req.session.woot).to.be('a');
        expect(req.session.likes).to.eql(['ferrets']);
        res.sendStatus(200);
      });

      request(app)
      .get('/')
      .expect(200)
      .end(function(err, res){
        if (err) return done(err);
        var cookie = res.headers['set-cookie'][0].split(';')[0];
        
        request(app)
        .get('/2')
        .set('Cookie', cookie)
        .end(function(err){
          if (err) return done(err);

          var cookie = res.headers['set-cookie'][0].split(';')[0];

          request(app)
          .get('/3')
          .set('Cookie', cookie)
          .expect(200, done);
        });
      });
    });
  });
});


index.js
"use strict";

/**
 * Module dependencies.
 */

const Expose = require('./expose');
const MyDBObserver = require('mydb-observer');

/**
 * `mydb-expose` middleware.
 *
 * @param {mongodb.Db} database to store sessions in
 * @param {Object} options
 * @api public
 */

function mydbExpose(db, options) {
  if ('undefined' === typeof db) {
    throw new Error('mydb-expose: No mongodb instance specified.')
  }

  options = options || {};
  options.session = options.session || {};

  // sessions collection
  // XXX: move session logic into `mydb-session`  
  let sessions = db.collection(options.session.collection || 'sessions');

  let observer;
  if (options.redis) {
    // make sure session updates are published to mydb
    observer = new MyDBObserver(options.redis);
    observer.observe(sessions);
  }

  sessions.ensureIndex('sid');

  // session fields to expose
  let sessionExpose = options.session.expose || { 'sid': 0 };

  // secret
  let secret = options.secret || 'youareagoodmydbcracker';

  // url
  let url = options.url;

  if (!url) {
    throw new Error('mydb-expose: Missing `url` (mydb server) option.');
  }

  if (typeof url == 'string') {
    let _url = url;
    url = function(){
      return _url;
    };
  }

  // create middleware
  return function (req, res, next) {
    let expose = new Expose(url, secret, sessions, sessionExpose);
    expose.middleware(req, res, next);
  };
}

/**
 * Module exports.
 */

module.exports = mydbExpose;

session.js
"use strict";

/**
 * Module dependencies.
 */

var debug = require('debug')('mydb-expose:session');

/**
 * Module exports.
 */

module.exports = Session;

/**
 * Noop.
 */

function noop(){}

/**
 * Session.
 *
 * @param {monk.Collections} collection
 * @param {ServerRequest} request
 * @api public
 */

function Session(col, req){
  // prefix to minimize collisions with keys
  this.$col = col;
  this.$req = req;
  this.$qry = {};
  this.$keys = [];

  // copy properties from the original session
  for (var i in req.session){
    if ('_csrf' == i) continue;
    if (!this[i]) this[i] = req.session[i];
  }
}

/**
 * Checks for dirty state.
 *
 * @return {Boolean}
 * @api public
 */

Session.prototype.$dirty = function(){
  return Object.keys(this.$qry).length > 0;
};

/**
 * Buffers a given query.
 *
 * @param {String} operation
 * @param {Object} operation obj
 * @api private
 */

Session.prototype.$buffer = function(op, obj){
  this.$qry[op] = this.$qry[op] || {};
  for (var i in obj) this.$qry[op][i] = obj[i];
};

/**
 * Populates the object.
 *
 * @param {Function} callback
 * @api private
 */

Session.prototype.reload = function(fn){
  fn = fn || noop;
  var self = this;
  var qry = { sid: this.$req.originalSession.id };
  var set = {
    sid: this.$req.originalSession.id,
    active_at: new Date,
    user_agent: this.$req.headers['user-agent'],
    state: 'online'
  };
  var opts = { upsert: true };
  debug('reload %s', qry.sid);
  this.$col.findAndModify(qry, [['sid', 1]], { $set: set }, opts, function(err, obj){
    if (err) return fn(err);
    if (obj.lastErrorObject && obj.lastErrorObject.upserted) {
      set._id = obj.lastErrorObject.upserted;
      obj = set;
    } else {
      obj = obj.value;
    }
    var keys = self.$keys;
    for (var i = 0; i < keys.length; i++) delete self[keys[i]];
    for (var i in obj) self[i] = obj[i];
    self.$keys = Object.keys(obj);
    debug('reloaded with %j', obj);
    fn(null);
  });
};

/**
 * Saves the document by performing the buffered operations.
 *
 * @api private
 */

Session.prototype.save = function(fn){
  fn = fn || noop;
  if (!this._id) return fn(new Error('Session not properly loaded'));
  debug('saving %j', this.$qry);
  this.$query(this.$qry, fn);
};

/**
 * Regenerates the original session.
 *
 * @param {Function} callback
 * @api public
 */

Session.prototype.regenerate = function(fn){
  fn = fn || noop;
  var self = this;
  this.$qry = {};
  debug('regenerating');
  this.$req.originalSession.regenerate(function(err){
    if (err) return fn(err);
    if (self !== self.$req.session) {
      // req.session has been replaced, restore it
      self.$req.originalSession = self.$req.session;
      self.$req.session = self;
    }
    // make sure the reference to the new cookie object is copied over
    self.cookie = self.$req.originalSession.cookie;
    self.reload(fn);
  });
  return this;
};

/**
 * Runs a query directly.
 *
 * @param {Object} query
 * @param {Function} callback
 * @api private
 */

Session.prototype.$query = function(qry, fn){
  if (!this._id) return fn(new Error('Session not properly loaded'));
  this.$col.update({ _id: this._id }, qry, fn);
};

/**
 * Sets `key` to `val`.
 *
 * @param {String} key
 * @param {Object} value
 * @param {Function} optional, callback
 * @api public
 */

Session.prototype.set = function(key, val, fn){
  var op = {};
  op[key] = val;

  if (fn) {
    debug('set %s : %j', key, val);
    this.$query({ $set: op }, fn);
  } else {
    debug('buffered set %s : %j', key, val);
    this.$buffer('$set', op);
  }

  return this;
};

/**
 * Pops `key`.
 *
 * @param {String} key
 * @param {Function} optional, callback
 * @api public
 */

Session.prototype.pop = function(key, fn){
  var op = {};
  op[key] = 1;

  if (fn) {
    debug('pop %s', key);
    this.$query({ $pop: op }, fn);
  } else {
    debug('buffered pop %s', key);
    this.$buffer('$pop', op);
  }

  return this;
};

/**
 * Shifts `key`.
 *
 * @param {String} key
 * @param {Function} optional, callback
 * @api public
 */

Session.prototype.shift = function(key, fn){
  var op = {};
  op[key] = -1;

  if (fn) {
    debug('shift %s', op);
    this.$query({ $pop: op }, fn);
  } else {
    debug('buffered shift %s', op);
    this.$buffer('$pop', op);
  }

  return this;
};

/**
 * Unsets `key`.
 *
 * @param {String} key to unset
 * @Param {Function} callback
 * @api public
 */

Session.prototype.unset = function(key, fn){
  var op = {};
  op[key] = 1;

  if (fn) {
    debug('unset %s', key);
    this.$query({ $unset: op }, fn);
  } else {
    debug('buffered unset %s', key);
    this.$buffer('$unset', op);
  }

  return this;
};

/**
 * Renames `key` to `key2`.
 *
 * @param {String} key to be renamed
 * @param {String} new key name
 * @param {Functio} optional, callback
 * @api private
 */

Session.prototype.rename = function(key, key2, fn){
  var op = {};
  op[key] = key2;

  if (fn) {
    debug('rename %s to %s', key, key2);
    this.$query({ $rename: op }, fn);
  } else {
    debug('buffered rename %s to %s', key, key2);
    this.$buffer('$rename', op);
  }

  return this;
};

/**
 * Pushes `val` to `key`.
 *
 * @param {String} key
 * @param {Object} value to push
 * @param {Function} optional, callback
 * @api public
 */

Session.prototype.push = function(key, val, fn){
  var op = {};
  op[key] = val;

  if (fn) {
    debug('push %s %j', key, val);
    this.$query({ $push: op }, fn);
  } else {
    debug('buffered push %s %j', key, val);
    this.$buffer('$push', op);
  }

  return this;
};

/**
 * Pushes `val` to `key` (set).
 *
 * @param {String} key
 * @param {Object} value to push
 * @param {Function} optional, callback
 * @api public
 */

Session.prototype.addToSet = function(key, val, fn){
  var op = {};
  op[key] = val;

  if (fn) {
    debug('add to set %s %j', key, val);
    this.$query({ $addToSet: op }, fn);
  } else {
    debug('buffered add to set %s %j', key, val);
    this.$buffer('$addToSet', op);
  }

  return this;
};

/**
 * Pushes `vals` to `key`.
 *
 * @param {String} key
 * @param {Array} value to push
 * @param {Function} optional, callback
 * @api public
 */

Session.prototype.pushAll = function(key, vals, fn){
  var op = {};
  op[key] = vals;

  if (fn) {
    debug('push all %s %j', key, vals);
    this.$query({ $pushAll: op }, fn);
  } else {
    debug('buffered push all %s %j', key, vals);
    this.$buffer('$pushAll', op);
  }

  return this;
};

/**
 * Pulls `val` from `key`.
 *
 * @param {String} key
 * @param {Object} value to push
 * @param {Function} optional, callback
 * @api public
 */

Session.prototype.pull = function(key, val, fn){
  var op = {};
  op[key] = val;

  if (fn) {
    debug('pull %s %j', key, val);
    this.$query({ $pull: op }, fn);
  } else {
    debug('buffered pull %s %j', key, val);
    this.$buffer('$pull', op);
  }

  return this;
};

/**
 * Pulls `vals` from `key`.
 *
 * @param {String} key
 * @param {Array} values to push
 * @param {Function} optional, callback
 * @api public
 */

Session.prototype.pullAll = function(key, vals, fn){
  var op = {};
  op[key] = vals;

  if (fn) {
    debug('pull all %s %j', key, vals);
    this.$query({ $pullAll: op }, fn);
  } else {
    debug('buffered pull all %s %j', key, vals);
    this.$buffer('$pullAll', op);
  }

  return this;
};

/**
 * Return JSON representation.
 *
 * @return {Object}
 * @api public
 */

Session.prototype.toJSON = function(){
  var obj = {};

  for (var key in this) {
    if (!this.hasOwnProperty(key)) continue;
    if ('$' == key[0]) continue;
    obj[key] = this[key];
  }

  return obj;
};


expose.js
"use strict";

/**
 * Module dependencies.
 */

const qs = require('querystring');
const url = require('url');
const uid = require('uid2');
const http = require('http');
const crypto = require('crypto');
const request = require('superagent');
const Session = require('./session');
const debug = require('debug')('mydb-expose');

/**
 * Module exports.
 */

module.exports = Expose;

/**
 * Expose constructor.
 *
 * @param {Function} url getter
 * @param {String} mydb secret
 * @param {mongodb.Collection} sessions collection
 * @param {String|Array|Object} session fields to expose
 * @api public
 */

function Expose(url, secret, sessions, expose){
  this.url = url;
  this.secret = secret;
  this.sessions = sessions;
  this.sessionExpose = expose;
}

/**
 * Returns the overriden `end`.
 *
 * @return {Function} end
 * @api private
 */

Expose.prototype.end = function(){
  var req = this.req;
  var res = this.res;
  var end = this.res.end;
  var next = this.next;

  return function(data, encoding){
    res.end = end;

    if (req.session && req.session.$dirty()) {
      req.session.save(done);
    } else {
      done();
    }

    function done(err){
      if (err) return next(err);
      if (req.originalSession) {
        debug('restoring original session');
        req.session = req.originalSession;
      }
      res.end(data, encoding);
    }
  };
};

/**
 * Returns the overriden `send`.
 *
 * @return {Function} send
 * @api private
 */

Expose.prototype.send = function(){
  var req = this.req;
  var res = this.res;
  var subscribe = req.subscribe;
  var send = res.send;
  var next = req.next;
  var self = this;

  return function (data, fields) {
    res.send = send;
    
    if ('function' === typeof data.toArray) {
      data = data.toArray();
    }

    if (data instanceof Promise) {
      debug('handling res#send for mongodb promise');
      if (req.get('X-MyDB-SocketId')) {  
        data.then(doc => {
          if (!doc || !doc._id) return (res.sendStatus || res.send).call(res, 404);
          debug('mydb - subscribing');
          subscribe(doc._id, fields || Object.keys(doc), function (err, id) {
            if (err) return next(err);
            if (id == req.get('X-MyDB-Id')) {
              debug('subscription id matches one provided by client');
              (res.sendStatus || res.send).call(res, 304);
            } else {
              debug('sending new subscription with document');
              res.set('X-MyDB-Id', id);
              res.send(doc);
            }
          });
        })
        .catch(next);
      } else {
        debug('no mydb - not subscribing');
        data.then(doc => {
          if (!doc) return (res.sendStatus || res.send).call(res, 404);
          res.send(doc);
        }).catch(next);
      }
    } else {
      res.send.apply(res, arguments);
    }
  };
};

/**
 * Fetches a document from a promise.
 *
 * @return {Function} send
 * @api private
 */

Expose.prototype.subscribe = function(){
  var req = this.req;
  var res = this.res;
  var send = res.send;
  var next = req.next;
  var self = this;

  /**
   * Subscribe.
   *
   * @param {String|ObjectId|Promise} data
   * @param {String|Array|Object} fields (optional)
   * @param {Function} callback (optional)
   * @api public
   */

  return function subscribe(data, fields, fn){
    if ('function' == typeof fields) {
      fn = fields;
      fields = null;
    }

    if (data instanceof Promise) {
      data.then(function(doc){
        if (!doc || !doc._id) return fn(new Error('Not found'));
        to(doc._id, fields || data.opts.fields);
      }).catch(fn);
    } else {
      to(data);
    }

    function to(id){
      id = id.toString();
      self.createSubscription(req.mydb_socketid, id, fields, fn);
    }
  };
};

/**
 * Subscribes to the given document.
 *
 * @param {String} socketid
 * @param {ObjectId|String} doc oid
 * @Param {String} socketid
 * @param {Object|String|Array} fields
 * @param {Function} callback
 * @api private
 */

Expose.prototype.createSubscription = function(socketid, id, fields, fn){
  fields = fields || {};
  var qry = {};

  // document id
  qry.document_id = id;

  // document fields
  fields = toFields(fields);
  if (Object.keys(fields).length) qry.fields = fields;

  // client sid
  qry.socket_id = socketid;

  // subscription id is a hash of fields/oid
  var ffs = JSON.stringify(order(fields || {}));
  var sid = qry.subscription_id = md5(socketid + '.' + id + '.' + ffs);

  // publish
  var data = JSON.stringify(qry);
  var start = Date.now();

  var url = this.url;

  // allow for a function getter
  if ('function' == typeof url) {
    url = url.call(this);
  }

  if ('object' == typeof url) {
    // using mydb object directly (in-proc)
    url.subscribe(qry, function(err){
      if (err) return fn(err);
      fn(null, sid);
    });
  } else {
    // post to mydb API
    var req = request
    .post(url + '/mydb/subscribe')
    .set('Content-Type', 'application/json')
    .set('X-MyDB-Signature', sign(data, this.secret))
    .send(data)
    .end(function(err, res){
      if (err) {
        debug('socket error');
        return fn(err);
      }

      if (res.error) {
        debug('subscription error %j', res.error);
        return fn(res.error);
      }

      debug('mydb subscribe took %d', Date.now() - start);
      fn(null, sid);
    });
  }
};

/**
 * Returns the overriden `session`.
 *
 * @return {MyDBSession} session object
 */

Expose.prototype.sess = function(){
  var session = new Session(this.sessions, this.req);
  this.req.originalSession = this.req.session;
  return session;
};

/**
 * Connect middleware.
 *
 * @param {ServerRequest} request
 * @param {ServerResponse} response
 * @return {Function} middleware
 * @api private
 */

Expose.prototype.middleware = function expose(req, res, next){
  // prevent double middleware
  if (req.mydb) {
    debug('skipping - mydb-expose already mounted');
    return next();
  }

  var self = this;

  // setup shortcut to instance
  req.mydb = res.mydb = this;

  // keep track of req and res objects
  this.req = req;
  this.res = res;
  this.next = next;

  // setup overrides
  res.subscribe = req.subscribe = this.subscribe();
  res.end = this.end();
  res.send = this.send();

  // generate a socketid if one is not set
  req.mydb_socketid = req.get('X-MyDB-SocketId');
  if (!req.mydb_socketid) {
    debug('creating new socketid');
    uid(12, function(err, id) {
      if (err) return next(err);
      req.mydb_socketid = id;
      done();
    });
  } else {
    debug('using socketid provided with request');
    done();
  }

  function done(){
    if (req.session) {
      // session object
      req.session = self.sess();

      // populates the session and moves on
      req.session.reload(function(err){
        if (err) return next(err);
        self.routes(next);
      });
    } else {
      debug('no session - skipping');
      next();
    }
  }
};

/**
 * Introduce mydb-expose routes.
 *
 * @param {Function} next
 * @api private
 */

Expose.prototype.routes = function(next){
  if (/^\/session\/?(\?.*)?$/.test(this.req.url) && 'GET' == this.req.method) {
    var col = this.sessions;
    var sid = this.req.session._id;
    var pro = col.findOne({ _id: sid }, { fields: this.sessionExpose });
    this.res.send(pro);
  } else {
    next();
  }
};

/**
 * Returns the connect middleware.
 *
 * @api private
 */

Expose.prototype.fn = function(){
  return this.middleware.bind(this);
};

/**
 * MD5 helper.
 *
 * @param {String} text
 * @api private
 */

function md5(text){
  return crypto.createHash('md5')
  .update(text)
  .digest('hex');
}

/**
 * Orders an object.
 *
 * @api private
 */

function order(o){
   var a = [], i;

   for (i in o) {
     if (o.hasOwnProperty(i)) {
       a.push([i,o[i]]);
     }
   }

   a.sort(function(a,b){
     return a[0] > b[0] ? 1 : -1;
   });

   return a;
}

/**
 * Fields argument helper.
 *
 * @param {String|Array|Object} fields
 * @api private
 */

function toFields(obj) {
  if (!Array.isArray(obj) && 'object' == typeof obj) {
    return obj;
  }

  var fields = {};
  obj = 'string' == typeof obj ? obj.split(' ') : (obj || []);

  for (var i = 0, l = obj.length; i < l; i++) {
    if ('-' == obj[i][0]) {
      fields[obj[i].substr(1)] = 0;
    } else {
      fields[obj[i]] = 1;
    }
  }

  return fields;
}

/**
 * HMac signing helper.
 *
 * @param {String} data
 * @param {String} secret
 * @api private
 */

function sign(data, secret){
  return crypto
  .createHmac('sha1', secret)
  .update(data)
  .digest('hex');
}
